fem/bilininteg.cpp

```diff
@@ -1688,93 +1688,36 @@ void VectorFEMassIntegrator::AssembleElementMatrix(
    ElementTransformation &Trans,
    DenseMatrix &elmat)
 {
-   int dof = el.GetDof();
-   int spaceDim = Trans.GetSpaceDim();
-
-   double w;
-
-#ifdef MFEM_THREAD_SAFE
-   Vector D(VQ ? VQ->GetVDim() : 0);
-   DenseMatrix trial_vshape(dof, spaceDim);
-   DenseMatrix K(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
-#else
-   trial_vshape.SetSize(dof, spaceDim);
-   D.SetSize(VQ ? VQ->GetVDim() : 0);
-   K.SetSize(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
-#endif
-   DenseMatrix tmp(trial_vshape.Height(), K.Width());
-
-   elmat.SetSize(dof);
-   elmat = 0.0;
-
-   const IntegrationRule *ir = IntRule;
-   if (ir == NULL)
-   {
-      // int order = 2 * el.GetOrder();
-      int order = Trans.OrderW() + 2 * el.GetOrder();
-      ir = &IntRules.Get(el.GetGeomType(), order);
-   }
-
-   for (int i = 0; i < ir->GetNPoints(); i++)
-   {
-      const IntegrationPoint &ip = ir->IntPoint(i);
-
-      Trans.SetIntPoint (&ip);
-
-      el.CalcVShape(Trans, trial_vshape);
-
-      w = ip.weight * Trans.Weight();
-      if (MQ)
-      {
-         MQ->Eval(K, Trans, ip);
-         K *= w;
-         Mult(trial_vshape,K,tmp);
-         AddMultABt(tmp,trial_vshape,elmat);
-      }
-      else if (VQ)
-      {
-         VQ->Eval(D, Trans, ip);
-         D *= w;
-         AddMultADAt(trial_vshape, D, elmat);
-      }
-      else
-      {
-         if (Q)
-         {
-            w *= Q -> Eval (Trans, ip);
-         }
-         AddMult_a_AAt (w, trial_vshape, elmat);
-      }
-   }
+   AssembleElementMatrix2(el, el, Trans, elmat);
 }
 
 void VectorFEMassIntegrator::AssembleElementMatrix2(
    const FiniteElement &trial_fe, const FiniteElement &test_fe,
    ElementTransformation &Trans, DenseMatrix &elmat)
 {
-   if ( test_fe.GetRangeType() == FiniteElement::SCALAR && VQ )
+   if ( test_fe.GetRangeType() == FiniteElement::SCALAR 
+             && trial_fe.GetRangeType() == FiniteElement::VECTOR )
    {
       // assume test_fe is scalar FE and trial_fe is vector FE
       int dim  = test_fe.GetDim();
       int trial_dof = trial_fe.GetDof();
       int test_dof = test_fe.GetDof();
       double w;
-
-      if (MQ)
-         mfem_error("VectorFEMassIntegrator::AssembleElementMatrix2(...)\n"
-                    "   is not implemented for tensor materials");
+      double Kv;
 
 #ifdef MFEM_THREAD_SAFE
       DenseMatrix trial_vshape(trial_dof, dim);
       Vector shape(test_dof);
-      Vector D(dim);
+      Vector D(VQ ? VQ->GetVDim() : 0);
+      DenseMatrix K(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
 #else
       trial_vshape.SetSize(trial_dof, dim);
       shape.SetSize(test_dof);
-      D.SetSize(dim);
+      D.SetSize(VQ ? VQ->GetVDim() : 0);
+      K.SetSize(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
 #endif
 
-      elmat.SetSize (test_dof, trial_dof);
```

Comment:
H1 is used to discretize each of the vector components, and so, elmat has to have size (dim*test_dof, trial_dof) in the case of SCALAR-VECTOR mix of basis.

Reply:
This was actually an old bug.

Reply:
Actually, I believe this was not a bug but an undocumented feature.  The code actually assumed `vdim=1` in the scalar field and computed the bilinear form `(v.w, u)` where `v` is the vector coefficient, `w` is the Nedelec or RT basis function, and `u` is the scalar basis function.  It is confusing to have this operator here and I could not find any examples or mini apps that use it so it is probably best to remove this (as this PR does).  Incidentally, this confusing functionality is now available, and documented, in `MixedDotProductIntegrator` if anyone needs it.